### PR TITLE
Uncrosscompiling: Basic build system, step 2 (take 2)

### PR DESCRIPTION
**Pull request by boricj**
_What happens when we don't cross-compile SerenityOS?_

Same basic idea as #6781, except this time it's hopefully more sustainable. Lagom is reimagined as a proper host build of the entire Serenity userland (that is, not cross-compiled), except everything is statically linked ~~out of laziness~~ in order to keep this PR at a sane size. This means that `-DBUILD_LAGOM=On` no longer cross-compiles a working SerenityOS system and should therefore be done in a separate build directory than `Build/i686`.

The SerenityOS system should still be fully buildable and working, although I haven't tested it thoroughly. What most definitively won't work is the CI because I did not (yet) hook everything together, hence the draft status. Also, the Lagom build requires continuing the build even on errors because the userland isn't actually ported.

I still stand by my conclusion in #6781 that the current build system has a lot of issues that were papered over. Hopefully I can nudge things one step at a time and also inspire people to work on Serenity's userland portability by exposing all of it to foreign systems.

### CODE REVIEW

**Comment by boricj** [comment]
So I'm mostly back to where this was previously, with 99 utilities building on 64-bit Linux after some `#ifdef`-ing of `pledge()/unveil()`. I'll refrain from fixing more portability things than that because this PR will become unmanageable otherwise, but a non-exhaustive list of failures include:
- Non-portable, SerenityOS-specific code (`#include <serenity.h>`, raw syscalls, `arc4random()` and `arc4random_uniform()` which should be abstracted by AK, `getsignalbyname()`) ;
- Non-compliant POSIX include paths (`#include <mman.h>`) ;
- Relying on non-standard header locations of POSIX definitions (`ARG_MAX`, `errno`) ;
- Non-64 bit clean code (cast to pointer from integer of different size, narrowing casts, printf formatters, `#include <sys/arch/i386/regs.h>`) ;
- Usage of LibELF, which is included inside LibC inside SerenityOS and thus not yet available elsewhere ;
- Issues with Serenity's existing portability code (`constexpr size_t buffer_size = PAGE_SIZE;` with `PAGE_SIZE` not being a constant) ;
- Const-correctness issues (`char* dot = strchr(d_string, '.')`) ;
- Tests utilities that should probably be moved elsewhere (`test_env`, `test_efault`, `test_unveil`).

Overall it's not quite as bad as one might think, but there is a _lot_ of work to do here. Carrying out these fixes would also improve SerenityOS's compatibility with ports and make some headway into a 64-bit SerenityOS port, so hopefully this PR would be of some use even if it ultimately fails or is rejected.

I also have a couple of services building, including WebServer. Alas, there seems to be a bug inside `IODevice::read_all()` when using a TCP socket on Linux as this function never returns, so it's not actually working. However, if running serenityos.org on SerenityOS is currently out of reach, running Serenity's web server on Linux might be doable with a little bit of work (#792).

I haven't bothered with applications and applets, because without a working port of WindowServer they won't be of much use. I also haven't bothered with GitHub actions, because I don't know if this approach is acceptable in the first place.

**Comment by gunnarbeutner** [comment]
FWIW WebServer shouldn't be using `read_all()` at all, see #7064. :)

File: Userland/Libraries/LibThread/Lock.h [hunk 1]
```diff
@@ -24,7 +27,22 @@ class Lock {
     void unlock();
 
 private:
-    Atomic<pid_t> m_holder { 0 };
+#ifdef __serenity__
+    using ThreadID = int;
+
+    ALWAYS_INLINE static ThreadID self()
+    {
+        return gettid();
```
Comment (gunnarbeutner) [inline Userland/Libraries/LibThread/Lock.h:35]: Maybe we should just use `pthread_t` and `pthread_self()` for SerenityOS. It's pretty much an alias for `gettid()` anyway.

File: Userland/Libraries/LibCore/CMakeLists.txt [hunk 2]
```diff
@@ -38,3 +38,7 @@ endif()
 
 serenity_lib(LibCore core)
 target_link_libraries(LibCore ${LIBRARY_LIBC} ${LIBRARY_LIBCRYPT})
+
+if(BUILD_LAGOM)
+    target_link_libraries(LibCore ${LIBRARY_LIBPTHREAD})
```
Comment (gunnarbeutner) [inline Userland/Libraries/LibCore/CMakeLists.txt:43]: Why is that necessary and should we perhaps also do it for SerenityOS?

File: Userland/Utilities/base64.cpp [hunk 3]
```diff
@@ -14,10 +14,12 @@
 
 int main(int argc, char** argv)
 {
+#ifdef __serenity__
```
Comment (gunnarbeutner) [inline Userland/Utilities/base64.cpp:17]: I'm not sure I like those `#ifdef`s. Maybe we should add a new library for compatibility functions and just link all targets against that.

GCC supports `#include_next` so we can "patch" the host's `<unistd.h>` header to make `pledge()`, `unveil()`, etc. available.

**Review by gunnarbeutner — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCore/CMakeLists.txt (hunk 2 above)
Comment (boricj) [reply Userland/Libraries/LibCore/CMakeLists.txt:43]: IIRC this comes from AK's StackInfo, which has a couple of calls to pthread on Linux.

File: Userland/Utilities/base64.cpp (hunk 3 above)
Comment (boricj) [reply Userland/Utilities/base64.cpp:17]: The `pledge()`/`unveil()` stuff isn't really portable to anything but SerenityOS, even for OpenBSD I seriously doubt that the pledges are actually compatible across the two operating systems.

NetBSD has a libcompat to deal with the portability of host tools. I'm not sure if we really should just stub out `pledge()`/`unveil()` and that could pose problems for OpenBSD too.
